### ISSUE TITLE
Sort based on resolved values

### ISSUE THREAD

**Issue by skosch** [bebraw/reactabular#160]
As mentioned in the original post #157 – `sort` should take four parameters: `data`, `columns`, `sortingColumns` (i.e. their indices), and `sort` (i.e. `orderBy`). The actual sorting function should act on the resolved values, not simply whatever `get(row, property)` returns.

The data operations would then look like
```
resolve/flatten -> sort -> apply order to original data ----> resolve -> format -> render
\------------------------------------------------------/      \-------------------------/
                       sorter                                            Table
```
or, alternatively, if sorting could be integrated into the table somehow, we'd avoid resolving everything twice (though I don't know if that can be done in a sane way):
```
resolve -> sort -> format -> render
\--------------------------------/
                Table
```

**Comment by bebraw** [comment]
I think the cheapest fix would look like this:

```diff
diff --git a/src/sort.js b/src/sort.js
index ecb73a8..d7a2a31 100644
--- a/src/sort.js
+++ b/src/sort.js
@@ -85,9 +85,10 @@ const sorter = ({ data, sortingColumns, sort }) => {
   sortingColumns.forEach((column) => {
     propertyList.push(row => {
       const value = get(row, column.property);
+      const formatter = column.cell.resolve || (a => a);
 
       if (value && value.toLowerCase) {
-        return value.toLowerCase();
+        return formatter(value).toLowerCase();
       }
 
       return value;
```

It's literally the same idea as for search now and would solve the problem I think.

> or, alternatively, if sorting could be integrated into the table somehow, we'd avoid resolving everything twice (though I don't know if that can be done in a sane way):

The current solution isn't very efficient, I agree.

I think what we actually want is the concept of pipelines. The full example has code like this

```javascript
let d = search.multipleColumns({ data, columns, query });

d = sort.sorter({ data: d, sortingColumns, sort: orderBy });

const paginated = paginate(d, pagination);
const pages = Math.ceil(d.length / Math.max(
  isNaN(pagination.perPage) ? 1 : pagination.perPage, 1)
);

// paginated.data goes to the Table
```

The problem is that each step has to resolve separately. Instead it might be fun to be able to write something like this:

```javascript
const d = pipeline(
  resolve(),
  search.multipleColumns({ columns, query }),
  sort.sorter({ sortingColumns, sort: orderBy }),
  paginate(pagination)
)(data);

// d goes to Table
```

That would look saner to me. Maybe `pipeline` should resolve automatically. Now I just thought about it in generic terms.

What do you think?

**Comment by skosch** [comment]
Well that would take some significant refactoring of things, wouldn't it ... but if you're up for it, I for one would be thrilled! :100:  Being able to pipe those transforms together would be the gold standard of simplicity and elegance :)

**Comment by bebraw** [comment]
Yeah, that's why I have those tests around. 👍 

I'll do the first change as well.

**Comment by bebraw** [comment]
There's a working solution now. Essentially I resolve at `sorter`.

Initially I tried writing a separate resolution pass but that got messy fast as I had to go through the whole data and resolve each field. Even with that editing proved problematic as it relies on **non-resolved** values (think dropdowns etc.).

I don't know what would be the ideal way to solve this. The current is a starting point but you end up resolving a lot.
